biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Indexer.java

```diff
@@ -131,10 +138,29 @@ private Repository readRepository() throws Exception {
 		List<Artifact> artifacts = p2.getArtifacts();
 		List<Promise<Resource>> fetched = new ArrayList<>(artifacts.size());
 		Set<URI> visited = new HashSet<>(artifacts.size());
+		Set<String> visitedHash = new HashSet<>(artifacts.size());
+		Map<String,Resource> existing = new HashMap<>();
+
+		if (getBridge() != null) {
+			Requirement requirement = new RequirementBuilder(P2_CAPABILITY_NAMESPACE).addAttribute(MD5_ATTRIBUTE, "*")
+					.buildSyntheticRequirement();
+			for (Capability capability : getBridge().getRepository().findProviders(asList(requirement)).get(
+					requirement)) {
+				Resource resource = capability.getResource();
+				String md5 = (String) capability.getAttributes().get(MD5_ATTRIBUTE);
+				existing.put(md5, resource);
```

Comment:
MD5 is known to be weak against deliberate collisions, but for accidental ones it is more than sufficient for this sort of application afaiu. I did consider a composite key like you suggested, I just wanted to minimise the change footprint ... but sure I suppose it makes sense to make it as safe as possible with the information we have since it doesn't really cost anything. Okay.